Add Map Demo

bipol opened the pull request:
This should add a "map-demo" instance so that we can start serving the little map app I've been playing around with on our infrastructure.  I'm not entirely sure if this is correct -- but the image should be serving a little duder on port 4000 if I did it right, which is probably unlikely!

Definitely need another set of eyes, here is the dockerfile for map-demo: https://github.com/smartatransit/map-demo/blob/master/Dockerfile
and I've pushed the repo here:
https://hub.docker.com/repository/docker/smartatransit/map-demo

bipol committed:
add map demo

xanderflood commented on map-demo.tf:
```diff
@@ -0,0 +1,20 @@
+
+  traefik_network_name = docker_network.traefik.id
+
+  //nginx publishes to port 80 on the host by default - to avoid a
```
If the module is set up correctly (🙏 🙏) I think you want to delete the `endpoint_spec` here and instead change `port = 80` to `port = 4000` above.

Explanation that you don't need to read: this is just there because (for some frustrating reason) the nginx image is designed to try to grab port 80 on the _host_, which is bad because (1) traefik needs that port so there'll be a conflict, and (2) we don't want service containers to own host ports, we just want traefik to contact them internally. There doesn't seem to be any way to tell nginx to **not** grab a host port, but we can at least tell it to grab one other than port 80 so that it won't conflict with traefik and crash. I picked some random numbers like 23893 because those are behind the firewall anyways.

Your image is based on `node:10` which doesn't try to grab a host port in the first place, it just gives you the option of exposing the container in whichever way you feel is appropriate. All you need to do is set the `port` variable to 4000, which will in turn populate the `traefik.service.map-demo.port` label, which tells traefik that when it forwards a request to containers for this service, it should contact them on port 4000.

xanderflood reviewed (approved):
In your `Dockerfile`, I think you want to drop the `EXPOSE` directive. By default, all your containers ports will be available from within the docker network - I _think_ you only need extra directives like that if you want to publish to a host port, which isn't necessary here.

However, leaving the `EXPOSE` there shouldn't cause any issues, (since you aren't trying to `EXPOSE 80` like nginx does) so if you prefer you can go ahead and merge this as is to see if it works and change that later. I'm giving this a 👍  but I'm not 100% confident in my own module so let me know if anything isn't working properly and we can figure it out!

xanderflood commented:
WAIT I just looked at the `service` module and found a bug! I'm foolishly setting the loadbalancer `port` label to 80 right now and totally ignoring the `port` variable. If you go into `modules/service/main.tf` and change that label value to `var.port` then hopefully this will work 🙏

bipol committed:
change ports

xander gave some feedback on what values to change, as well as fixin a bug

bipol commented:
@xanderflood applied some of your changes -- it looks like because I'm not setting `endpoint_spec` anymore, and it is a `dynamic` field, the plan is failing?
```

Error: Missing map element

  on modules/service/main.tf line 52, in resource "docker_service" "service":
  52:       for_each = var.endpoint_spec.ports
    |----------------
    | var.endpoint_spec is empty map of dynamic
```

not sure how to fix this, but haven't spent any time poking at it!

xanderflood commented:
@bipol ahh that's interesting, it looks like my `default = {}` needs to be `default = { ports = [] }`. That is making me realize that this whole idea of "passing through service fields from outside the module" is a lot of work and what should probably be done instead is (1) delete the `endpoint_spec` variable entirely, (2) delete `nginxes.tf`, which should be the only place it's used.

My thinking now is that if in the future we need to declare endpoints/mount volumes etc, we should just not use the module in the first place, and the module is just here for the simplest case 🤷‍♀

bipol committed:
remove endpoint_spec

xander left a comment that removing the endpoint spec might be a good idea,
since passing these fields through a module is probably more customization than
we need for this module
